src/service/resource.js

```diff
@@ -560,6 +566,32 @@ export class Resource {
   }
 
   /**
+   * @return {!Promise} resolves when underlying element is built and within the
+   *    range specified by implementation's renderOutsideViewport
+   */
+  whenWithinRenderOutsideViewport() {
+    if (this.renderOutsideViewportPromise_) {
+      return this.renderOutsideViewportPromise_;
+    }
+    return this.renderOutsideViewportPromise_ = new Promise(resolver => {
+      this.renderOutsideViewportResolve_ = resolver;
+    });
+  }
+
+  /**
+   * @private resolves render outside viewport promise if one was created via
+   *    whenWithinRenderOutsideViewport.
+   */
+  resolveRenderOutsideViewport_() {
+    if (!this.renderOutsideViewportResolve_) {
```

Comment:
There should also be some tests for `resources.js` that check the behavior of this thing in isolation.  In particular, check that multiple calls in sequence Do The Right Thing.